### ISSUE TITLE
How to use inpaint/outpaint models?

### ISSUE THREAD

**Issue by WainWong** [Kosinkadink/ComfyUI-Advanced-ControlNet#133]
I noticed that you have added support for Controlnet++, thank you very much for your work!

But I would also like to ask, in the new controlnet union, it is claimed that inpaint is supported, but I don't know how to use this feature through your node, can you please take the time to answer my query?

**Comment by WainWong** [comment]
![workflow (69)](https://github.com/user-attachments/assets/e9610e0f-e838-4cad-ab3d-2a70cdbb1f06)
I'm currently executing inpaint like this, but the drawing comes out as a solid black image

**Comment by liushh39** [comment]
have you solved this problem?

**Comment by WainWong** [comment]
> [liushh39](/liushh39)

no

**Comment by jokero3answer** [comment]
![image](https://github.com/user-attachments/assets/a1a16b38-b715-4834-abae-f5d640eea37d)
You could try this.

**Comment by hben35096** [comment]
Me too @WainWong

**Comment by Kosinkadink** [comment]
The ControlNet++ inpaint/outpaint probably needs a special preprocessor for itself. From some light testing I just did, if you provide an unprocessed image in, it results something that looks like the colors are inverted, and if you provide an inverted image, it looks like some channels might be switched around.

I'll look into the diffusers code sometime soon(ish) to see what the input is expected to look like. In the meantime, here are the wonky results if you try to put in a normal image and an inverted image:
![image](https://github.com/user-attachments/assets/0c263d16-4cd1-4f59-ae06-438d61511944)

**Comment by hben35096** [comment]
![PixPin_2024-07-19_04-47-56](https://github.com/user-attachments/assets/942ce0e2-cdaf-46d9-af74-498ff06ad0ec)
It seems that an RGB of 0,0,0 would be considered a mask area

**Comment by WainWong** [comment]
> ![image](https://private-user-images.githubusercontent.com/130854108/349914256-a1a16b38-b715-4834-abae-f5d640eea37d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNTEzNzAsIm5iZiI6MTcyMTM1MTA3MCwicGF0aCI6Ii8xMzA4NTQxMDgvMzQ5OTE0MjU2LWExYTE2YjM4LWI3MTUtNDgzNC1hYmFlLWY1ZDY0MGVlYTM3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQwMTA0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYjJlNmY0NjU2MjgyMzgzODJkNjI2ZGM5NWU3YjQyN2Y1ODA5MTM5ZjI3Njg0NjE5YWY3MTUyNTU5ZGM3ZDNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xDJfe4wIri5rtI0j7O69l_CSkGHLXBGg6kGleD6zyOs) You could try this.



> ![PixPin_2024-07-19_04-47-56](https://private-user-images.githubusercontent.com/139383150/350139791-942ce0e2-cdaf-46d9-af74-498ff06ad0ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNTEzNzAsIm5iZiI6MTcyMTM1MTA3MCwicGF0aCI6Ii8xMzkzODMxNTAvMzUwMTM5NzkxLTk0MmNlMGUyLWNkYWYtNDZkOS1hZjc0LTQ5OGZmMDZhZDBlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQwMTA0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YTYzODc3NmUyOWRkZjgwMjE0ZGQ0MGY1NjE0Zjc5ZjQ1YjNjZjkxYjEyYTcyZDRmMWVmNzVhZjFlZjIzMmEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.I_JxMecQNb3tTy_dipbeYmIsfnKY6eqJZNArT_YqoYk) It seems that an RGB of 0,0,0 would be considered a mask area ![workflowcn3](https://private-user-images.githubusercontent.com/139383150/350140697-16f205dc-eea3-46b0-88c1-6c42810e4069.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNTEzNzAsIm5iZiI6MTcyMTM1MTA3MCwicGF0aCI6Ii8xMzkzODMxNTAvMzUwMTQwNjk3LTE2ZjIwNWRjLWVlYTMtNDZiMC04OGMxLTZjNDI4MTBlNDA2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQwMTA0MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYjcxZGE5Nzg1YzMwYjc5YTMyYmI5YzMzNzUyYzNhODllMWQyOTY2ZGY1MDY1NTZjNWRkMGNkNzBkNjU5YzBjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wdKynBU-bs3puZR8YxW15BASa9kF1k1PXEAvWVIKtm4)

Thanks, yes this works fine; but it looks like it's prone to getting redrawn if there are solid black values in the original image; hopefully there's a more elegant way to accomplish this

**Comment by jokero3answer** [comment]
this
![image](https://github.com/user-attachments/assets/d5f6084c-574c-42f4-b337-8557a8a459d7)

**Comment by hben35096** [comment]
You can also do this
![PixPin_2024-07-19_13-09-46](https://github.com/user-attachments/assets/576234f8-c0aa-4a8c-b3a0-a053361a82b2)

**Comment by 3blackbar** [comment]
OK this workarouind needs to go, its pretty weird to use workarounds like this and the only mode that can be used now is inpaint cause other union models suddenlty do not work cause of that mask conversion nodes.
This must be absolutely handled under the hood by comfy nodes, not by the user.Xinsir trained model with black region requirements so this must be taken care of when implementing his models into comfy nodes, just adding code to identify union modes is not enough, masking should work for black regions, so conversion must be done under the hood just for the union.

**Comment by Kosinkadink** [comment]
It's not much of a workaround, as having 'black' be the trigger for masked area is already unusual in itself. Normal inpaint controlnets expect -1 for where they should be masked, which is what the controlnet-aux Inpaint Preprocessor returns. The fact that OG controlnets use -1 instead of 0s for the mask is a blessing in that they sorta work even if you don't provide an explicit noise mask, as -1 would not normally be a value encountered by anything.

I'm pretty sure the original union code assumes that there is a noise mask applied in addition to the black RGB values in the mask itself, as just having 0,0,0 values is obviously not gonna work compared to the OG -1,-1,-1 values. The pipeline used in the diffusers code also applies the mask to the latents themselves, so you'll likely need to set up noise_mask (inpaint related stuff) to use inpaint/outpaint as intended. As for testing things out on my end right now, 1) I am busy with other things and 2) I personally don't use SDXL at all, so it's not something I am personally passionate about getting on ASAP compared to other things. If someone wants to look at the original code to confirm, please do so and let me know. The preprocessor for turning a mask into black pixels is dead simple, so I can add it this weekend, but the noise mask node already exists for inpainting in vanilla ComfyUI. Not sure if I'd want to make a 'combined' preprocessor that would simplify inpainting/outpainting using this controlnet or if it would have other consequences, but I can look into that at some point this weekend too.

My guess for the 0,0,0 values instead of -1,-1,-1 values in union is that since it is trying to be an 'all-in-one' controlnet, it would absolutely mess up all the training if just one of the input types happened to use negative values while everything else was scaled for [0, 1]. Even for other controlnet inputs, if there are negatives present in the cond image, something goes either to inf or Nan and ends up returning a black image, so it's a quirk of the model.

TL;DR: in addition to needing a special preprocess node to making mask make the pixels 0,0,0 in the input image, inpaint/outpaint type seems to require that you follow extra inpaint steps like setting a noise mask on the latents in a way that would match up with the black pixel mask applied to the controlnet inputs. I've included screenshots of the vanilla ComfyUI nodes that would include a noise_mask on the latents.
![image](https://github.com/user-attachments/assets/df7accd0-5eeb-4d6f-a5c8-68a3dffe0199)


@jokero3answer Hopefully this clears things up. Might be redundant to say after typing all this, but the black requirement has nothing to do with my code and is just the way promax 'repaint' mode was trained.

**Comment by 3blackbar** [comment]
Its Cnet++ so id epxect other unusual things, its AI,  the whole idea about inpainting is unusual, the tech is so new id not assume to even try standardize controlnet, unwillingness to expand will just force someone else to fork it and go own way.
I hate the black mask thing too but hey i want it to work so i just made it work for myself with crop and stitch.
Id make separate input for mask  just for xinsir inpaint and separate for image and forget about it move on.

**Comment by Kosinkadink** [comment]
There is no unwillingness to expand here. I've been supporting Advanced-ControlNet for a year now and I don't intend on stopping anytime soon.

From my understanding, the inpaint for union just needs a noise mask applied to the latents, which ComfyUI already supports with native nodes, so it can be tested. If there is more than that needed and there is a side by side comparison in the results to show it, please do let me know and we can work on having it be added in. I tagged this issue with "help needed" for a reason, I haven't had the chance to get into the weeds yet of ProMax-only types (tile and repaint), and I'm not sure if the existing preprocessors in control-aux (control-aux is not my repo btw) do what union/promax expects as input and if more needs to be done other than the noise mask. PRs are welcome too, even if just to share snippets of code to inspire the actual implementation.

**Comment by pcgamercc** [comment]
Hello @Kosinkadink can I bump in this issue with inpaint/outpaint in this same thread?
is this a file versioning issue?
thanks


```
Error occurred when executing ACN_ControlNet++LoaderAdvanced:

ControlNet++ model 'controlnet-union-sdxl-1.safetensors' does not support control_type 'inpaint/outpaint'.

  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\execution.py", line 317, in execute
    output_data, output_ui, has_subgraph = get_output_data(obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\execution.py", line 192, in get_output_data
    return_values = _map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\execution.py", line 169, in _map_node_over_list
    process_inputs(input_dict, i)
  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\execution.py", line 158, in process_inputs
    results.append(getattr(obj, func)(**inputs))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\custom_nodes\ComfyUI-Advanced-ControlNet\adv_control\nodes_plusplus.py", line 28, in load_controlnet_plusplus
    controlnet.verify_control_type(name, plus_input)
  File "C:\Users\GAMING PC\Documents\Ai_save\ComfyUI_windows_portable\ComfyUI\custom_nodes\ComfyUI-Advanced-ControlNet\adv_control\control_plusplus.py", line 248, in verify_control_type
    raise Exception(f"ControlNet++ model '{model_name}' does not support control_type '{pp_input.control_type}'.")
```

**Comment by Kosinkadink** [comment]
You'll need the Promax model (union was the initial model, Promax was basically the v2 model that added additional control types).

**Comment by pcgamercc** [comment]
Thank you, you were bang on right. the promax file version was the issue.

**Comment by Lannister34** [comment]
How to combine inpaint with other models like openpose?
